sql/analyzer/join_search.go

```diff
@@ -188,33 +184,86 @@ func permutations(a []int) (res [][]int) {
-	joinConds            []*joinCond
-	usedJoinCondsIndexes []int
-}
-
-func (js *joinSearchParams) copy() *joinSearchParams {
-	usedTableIndexesCopy := make([]int, len(js.usedTableIndexes))
-	copy(usedTableIndexesCopy, js.usedTableIndexes)
-	usedJoinCondIndexesCopy := make([]int, len(js.usedJoinCondsIndexes))
-	copy(usedJoinCondIndexesCopy, js.usedJoinCondsIndexes)
-	return &joinSearchParams{
-		tables:               js.tables,
-		usedTableIndexes:     usedTableIndexesCopy,
-		joinConds:            js.joinConds,
-		usedJoinCondsIndexes: usedJoinCondIndexesCopy,
+// visitJoinSearchNodes visits every possible joinSearchNode where the
+// in-order leaves are given by |tables|. If the callback returns
+// |false|, visits stop.
+func visitJoinSearchNodes(tables []string, cb func (n *joinSearchNode) bool) {
+	if len(tables) == 0 {
+		return
+	}
+	if len(tables) == 1 {
+		cb(&joinSearchNode{table: tables[0]})
+		return
+	}
+	var stop bool
+	for i := 1; i < len(tables) && !stop; i++ {
+		visitJoinSearchNodes(tables[:i], func (l *joinSearchNode) bool {
+			visitJoinSearchNodes(tables[i:len(tables)], func (r *joinSearchNode) bool {
+				if !cb(&joinSearchNode{left: l, right: r}) {
+					stop = true
+					return false
+				}
+				return true
+			})
+			return !stop
+		})
 	}
 }
 
-func (js *joinSearchParams) tableIndexUsed(i int) bool {
-	return indexOfInt(i, js.usedTableIndexes) >= 0
-}
-
-func (js *joinSearchParams) joinCondIndexUsed(i int) bool {
-	return indexOfInt(i, js.usedJoinCondsIndexes) >= 0
+// assignConditions attempts to assign the conditions in |conditions|
+// to the search tree in |root|, such that every condition is on an
+// internal node, and all of the trees referenced in the condition
+// appear in tables which are in the subtree of its internal node. If
+// it finds an assignment, leaves it in the |joinSearchNode.joinCond|
+// fields of the provided tree. Otherwise there is no valid assignment
+// and leaves the provided tree unmodified.
+func assignConditions(root *joinSearchNode, conditions []*joinCond) {
+	// A recursive helper which is going to assign conditions to
+	// subtrees, remove the assigned conditions from |conditions|
+	// and make a callback to |cb| for each such assignment that
+	// is found.
+	var helper func (n *joinSearchNode, cb func() bool) bool
+	helper = func (n *joinSearchNode, cb func() bool) bool {
+		if n.table != "" {
+			return cb()
+		}
+		// for each assignment of conditions to the left tree
+		return helper(n.left, func() bool {
+			// for each assignment of conditions to the right tree
+			return helper(n.right, func() bool {
+				tables := n.tableOrder()
+				// look at every remaining condition
+				for i := range conditions {
+					cond := conditions[i]
+					joinCondTables := findTables(cond.cond)
+					// if the condition only references tables in our subtree
+					if containsAll(joinCondTables, tables) {
+						n.joinCond = cond
+						copy(conditions[i:], conditions[i+1:])
```

Comment:
To remove a slice element, I find this to be clearer:

`a = append(a[:i], a[i+1:]...)`